File: 5/5-b.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type seed struct {
	id_start int
	id_end   int
}

type rule struct {
	dest_id_start int
	dest_id_end   int
	src_id_start  int
	src_id_end    int
}

type mapping struct {
	dest_name string
	val       []rule
}

type returns struct {
	id_start  int
	id_end    int
	dest_name string
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	reNumbers := regexp.MustCompile("^seeds: ([0-9 ]+)$")
	reMap := regexp.MustCompile("^(.*)-to-(.*) map:$")

	seeds := make([]seed, 0, 0)
	conversions := make(map[string]mapping)
	src_map := ""
	dst_map := ""
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			src_map = ""
			dst_map = ""
			continue
		} else if reNumbers.MatchString(line) {
			matches := reNumbers.FindStringSubmatch(line)
			splits := strings.Split(matches[1], " ")
			for i := 0; i < len(splits); i = i + 2 {
				v1 := splits[i]
				v2 := splits[i+1]
				i1, err := strconv.Atoi(v1)
				if err != nil {
					fmt.Println(err)
					return
				}
				i2, err := strconv.Atoi(v2)
				if err != nil {
					fmt.Println(err)
					return
				}
				aseed := seed{id_start: i1, id_end: i1 + i2 - 1}
				seeds = append(seeds, aseed)
			}
			continue
		} else if reMap.MatchString(line) {
			matches := reMap.FindStringSubmatch(line)
			src_map = matches[1]
			dst_map = matches[2]
			dst_mapping := mapping{dest_name: dst_map, val: make([]rule, 0, 0)}

			conversions[src_map] = dst_mapping
		} else {
			matches := strings.Split(line, " ")
			dst_start, err := strconv.Atoi(matches[0])
			if err != nil {
				fmt.Println(err)
				return
			}
			src_start, err := strconv.Atoi(matches[1])
			if err != nil {
				fmt.Println(err)
				return
			}
			count, err := strconv.Atoi(matches[2])
			if err != nil {
				fmt.Println(err)
				return
			}

			node := rule{
				dest_id_start: dst_start,
				dest_id_end:   dst_start + count - 1,
				src_id_start:  src_start,
				src_id_end:    src_start + count - 1,
			}

			mappingval := conversions[src_map]
			mappingval.val = append(mappingval.val, node)
			conversions[src_map] = mappingval
		}

	}

	locations := make([]returns, 0, 0)

	for _, aseed := range seeds {
		rets := get_location(aseed, conversions, "seed")
		locations = append(locations, rets...)
	}

	min := -1
	for _, location := range locations {
		if min == -1 || min > location.id_start {
			min = location.id_start
		}
	}
	fmt.Print("Minimum location: ")
	fmt.Println(min)
}

func get_location(aseed seed, conversion map[string]mapping, src_name string) []returns {
	dest := conversion[src_name].dest_name
	for _, rule := range conversion[src_name].val {
		range_start := rule.src_id_start
		range_end := rule.src_id_end

		delta := rule.src_id_start - rule.dest_id_start

		ranges := make([]seed, 0, 0)
		var check seed
		if aseed.id_start >= range_start && aseed.id_end <= range_end {
			check = aseed
		} else if aseed.id_start < range_start && aseed.id_end >= range_start && aseed.id_end <= range_end {
			// We start before the range
			ranges = append(ranges, seed{id_start: aseed.id_start, id_end: range_start - 1})
			check = seed{id_start: range_start, id_end: aseed.id_end}
		} else if aseed.id_start >= range_start && aseed.id_start <= range_end {
			// We start inside the ragne, but end outside it
			ranges = append(ranges, seed{id_start: range_end + 1, id_end: aseed.id_end})
			check = seed{id_start: aseed.id_start, id_end: range_end}
		} else if aseed.id_start < range_start && aseed.id_end > range_end {
			// We start before, end after
			ranges = append(ranges, seed{id_start: aseed.id_start, id_end: range_start - 1})
			ranges = append(ranges, seed{id_start: range_end + 1, id_end: aseed.id_end})
			check = seed{id_start: range_start, id_end: range_end}
		} else {
			continue
		}

		rets := make([]returns, 0, 0)
		for _, next_range := range ranges {
			rets = append(rets, get_location(next_range, conversion, src_name)...)
		}

		countdelta_start := check.id_start - range_start
		countdelta_end := check.id_end - range_start
		dst_id_start := range_start - delta + countdelta_start
		dst_id_end := range_start - delta + countdelta_end

		if dest == "location" {
			rets = append(rets, returns{id_start: dst_id_start, id_end: dst_id_end, dest_name: dest})
		} else {
			rets = append(rets, get_location(seed{id_start: dst_id_start, id_end: dst_id_end}, conversion, dest)...)
		}
		return rets
	}

	// Don't change the ID
	if dest == "location" {
		rets := make([]returns, 0, 0)
		rets = append(rets, returns{id_start: aseed.id_start, id_end: aseed.id_end, dest_name: dest})
		return rets
	}
	return get_location(aseed, conversion, dest)
}

File: 2023/5/5-b.pl
#!/usr/bin/perl

#
# Copyright (C) 2023 Joelle Maslak
# All Rights Reserved - See License
#

use JTM::Boilerplate 'script';
use List::Util qw(min);

MAIN: {
    my ( %maps,       @seeds );
    my ( $source_map, $dest_map );
    while ( my $line = <<>> ) {
        chomp($line);

        if ( $line eq "" ) {
            $source_map = undef;
            $dest_map   = undef;
            next;
        }
        if ( $line =~ m/^seeds: / ) {
            (@seeds) = ( split /\s+/, ( $line =~ m/^seeds: ([\d\s]+)$/ )[0] );
            next;
        } elsif ( $line =~ m/-to-.*map:$/ ) {
            ( $source_map, $dest_map ) = $line =~ m/^(.*)-to-(.*)\s+map:$/;
            if ( !exists $maps{$source_map} ) {
                $maps{$source_map} = {};
            }
            $maps{$source_map}{$dest_map} = [];
        } else {
            my ( $dest_start, $source_start, $count ) = split /\s+/, $line;
            push $maps{$source_map}{$dest_map}->@*,
              {
                source_start => $source_start,
                dest_start   => $dest_start,
                count        => $count,
              };
        }
    }

    my @locations;
    for ( my $i = 0; $i < scalar(@seeds); $i += 2 ) {
        my $seed_start = $seeds[$i];
        my $seed_end   = $seeds[$i] + $seeds[ $i + 1 ] - 1;
        my $rets       = get_location( $seed_start, $seed_end, \%maps, 'seed' );
        for my $ret ( $rets->@* ) {
            if ( $ret->[0] eq 'location' ) {
                push @locations, $ret->[1];
            }
        }
    }
    say "Minimum location: " . min(@locations);
}

sub get_location ( $source_id_start, $source_id_end, $maps, $source ) {
    # This is ugly and wrong, but we basically only ever have one
    # destinatio type for any given range, so the for loop with a return
    # at the end actually works.
    for my $dest ( keys $maps->{$source}->%* ) {
        for my $rule ( $maps->{$source}{$dest}->@* ) {
            my $range_start = $rule->{source_start};
            my $range_end   = $rule->{source_start} + $rule->{count} - 1;

            my $delta = $rule->{source_start} - $rule->{dest_start};
            my @ranges;
            my $check;
            if ( $source_id_start >= $range_start and $source_id_end <= $range_end ) {
                $check = [ $source_id_start, $source_id_end ];
            } elsif ( $source_id_start < $range_start
                and $source_id_end >= $range_start
                and $source_id_end <= $range_end )
            {
                # We start before range
                push @ranges, [ $source_id_start, $range_start - 1 ];
                $check = [ $range_start, $source_id_end ];
            } elsif ( $source_id_start >= $range_start and $source_id_start <= $range_end ) {
                # We start inside the range, but end outside it
                $check = [ $source_id_start, $range_end ];
                push @ranges, [ $range_end + 1, $source_id_end ];
            } elsif ( $source_id_start < $range_start and $source_id_end > $range_end ) {
                # We start before, end after
                $check = [ $range_start, $range_end ];
                push @ranges, [ $source_id_start, $range_start - 1 ];
                push @ranges, [ $range_end + 1, $source_id_end ];
            } else {
                next;
            }

            my @rets;
            for my $range (@ranges) {
                push @rets, get_location( $range->[0], $range->[1], $maps, $source )->@*;
            }

            my $countdelta_start = $check->[0] - $rule->{source_start};
            my $countdelta_end   = $check->[1] - $rule->{source_start};
            my $dest_id_start    = $rule->{source_start} - $delta + $countdelta_start;
            my $dest_id_end      = $rule->{source_start} - $delta + $countdelta_end;

            if ( $dest eq 'location' ) {
                push @rets, [ 'location', $dest_id_start, $dest_id_end ];
            } else {
                push @rets, get_location( $dest_id_start, $dest_id_end, $maps, $dest )->@*;
            }
            return \@rets;
        }

        # Don't change the ID.
        if ( $dest eq 'location' ) { return [ [ 'location', $source_id_start, $source_id_end ] ]; }
        return get_location( $source_id_start, $source_id_end, $maps, $dest );
    }
    return [];
}
